client/src/panels/panel_declarations/finances/transfer_payments/tp_by_region.js

```diff
@@ -65,183 +159,97 @@ class TPMap extends React.Component {
-        .max()
-        .value();
-      const color_scale_tp = d3.scaleLinear()
-        .domain([0, max_tp])
-        .range([0.2, 1]);
-      const largest_prov_tp = _.chain(current_year_data_tp)
-        .keys()
-        .maxBy( (prov) => current_year_data_tp[prov] )
-        .value();
-      const total_sum_tp = _.reduce(
-        current_year_data_tp,
-        (sum, value) => sum += value,
-        0
+      const transfer_payments_by_prov = transfer_payments_table.prov_code(
+        std_years,
+        subject.level === 'dept' && subject.id
       );
-      const percent_of_total_tp = current_year_data_tp[largest_prov_tp] / total_sum_tp;
-      const text_args_tp = {
-        largest_prov: provinces_with_article[largest_prov_tp].text,
-        largest_amount: formatter(current_year_data_tp[largest_prov_tp]),
-        total_sum: formatter(total_sum_tp),
-        percent_of_total: formats["percentage1_raw"](percent_of_total_tp),
-        subject: calculations.subject,
-        show_per_capita: false,
-      };
-
-      const should_tab_be_disabled = _.every(data_tppc, _.isEmpty);
-      
-      //TP PER CAPITA VERSION
-      const current_year_data_tppc = _.last(data_tppc);
-      
-      const max_tppc = _.chain(data_tppc)
-        .last()
-        .values()
-        .max()
+      const per_capita_by_prov = _.chain(transfer_payments_by_prov)
+        .omitBy( (values, prov_code) => _.isUndefined(population_data[prov_code]) )
+        .mapValues(
+          (transfer_payment_values, prov_code) => _.chain(transfer_payment_values)
+            .zipWith(
+              population_data[prov_code],
+              (transfer_payment, population) => transfer_payment/population
+            )
+            .value()
+        )
         .value();
-      const color_scale_tppc = d3.scaleLinear()
-        .domain([0, max_tppc])
-        .range([0.2, 1]);
 
-      const largest_prov_tppc = _.chain(current_year_data_tppc)
-        .keys()
-        .maxBy( (prov) => current_year_data_tppc[prov] )
-        .value();
-      const text_args_tppc = {
-        largest_prov: provinces_with_article[largest_prov_tppc].text,
-        largest_amount: formatter(current_year_data_tp[largest_prov_tppc]),
-        largest_per_capita: formatter(current_year_data_tppc[largest_prov_tppc]),
-        total_sum: formatter(total_sum_tp),
-        percent_of_total: formats["percentage1_raw"](percent_of_total_tp),
-        subject: calculations.subject,
-        show_per_capita: true,
-      };
+      const transfer_payment_data = group_prov_data_by_year(transfer_payments_by_prov);
+      const per_capita_data = group_prov_data_by_year(per_capita_by_prov);
+      
+      const text_args = get_text_args(subject, transfer_payment_data, per_capita_data);
 
       return (
         <StdPanel
           title={text_maker("tp_by_region_title")}
           {...{ footnotes, sources }}
         >
           <Col size={12} isText>
+            <TM k="tp_by_region_text" args={text_args} />
```

Comment:
Moved the panel text above the tabs. I found that even I couldn't be bothered to read the text carefully enough to tell what was changing in it when the tabs switched. Becomes sort of a word search when you know the text is changing in response to a click, but it's only changing by about 15%. Simplified the design AND the code to just have one static set of text, existing outside of the control of the tabs.

Reply:
Args were a little redundant, simplifying reduced clutter
👍